### DISCUSSION TITLE
gmail test mail timeout

### DISCUSSION THREAD

**Discussion by ghost** [pocketbase/pocketbase#458] [Q&A] [answered]
I still don't know if this is a bug. When trying to setup gmail as `smtp.gmail.com` the UI throws a `Test email send timeout.` error.

I've created an App password for mail in a google account and tested it using a linux mail client to see if the error was on Google's side. But it worked so I assume I might be reading something wrong or PB is not meant to be configured with gmail whatsoever.

Settings tried:

- Using the complete email address as username
- Using only the username without @gmail.com
- Using Auto(Start TLS)
- Using Always
- Changing the SMTP port

A weird thing I've noticed is that, even if I use a random username like idk, `asrfgvlksdnvjd` or something that might not exist the `Test email timeout` error still comes out.

How pb was deployed

PB runs on a systemd file on 0.0.0.0
Uses an SSL certificate on the Web UI
Reverse proxy to a subdomain,

**Comment by jstbyte** [comment]
> I still don't know if this is a bug. When trying to setup gmail as `smtp.gmail.com` the UI throws a `Test email send timeout.` error.
> 
> I've created an App password for mail in a google account and tested it using a linux mail client to see if the error was on Google's side. But it worked so I assume I might be reading something wrong or PB is not meant to be configured with gmail whatsoever.
> 
> Settings tried:
> 
> * Using the complete email address as username
> * Using only the username without @gmail.com
> * Using Auto(Start TLS)
> * Using Always
> * Changing the SMTP port
> 
> A weird thing I've noticed is that, even if I use a random username like idk, `asrfgvlksdnvjd` or something that might not exist the `Test email timeout` error still comes out.
> 
> How pb was deployed
> 
> PB runs on a systemd file on 0.0.0.0 Uses an SSL certificate on the Web UI Reverse proxy to a subdomain,
--------------------------------------------------------------------------------------------------------------------------------------------
Use Following :-

* smtp: smtp.gmail.com
* TLS port: 587
* User Name: example@gmail.com
* App Password: *********

**Reply by ghost** [reply]
It's still not working with both TLS auto and TLS always :(

**Reply by ghost** [reply]
![image](https://user-images.githubusercontent.com/24773698/190337498-f7d09f2d-f2fb-4e6f-98ae-a4de8a62b629.png)

(Mail edited :sweat_smile: )

**Reply by ganigeorgiev** [reply]
@VentGrey Do you have a firewall (eg. selinux) that could block the traffic for outgoing ports?

The "_Test email send timeout._" is thrown when the UI fails to receive a response from the test endpoint after 30s.

_I've tested it with a test google workspace account and Less secure apps enabled._

**Reply by ganigeorgiev** [reply]
@VentGrey As a side note, your issue is maybe caused also by the fact that the standard `net/smtp` is sending "localhost" for HELO/EHLO by default (see https://github.com/pocketbase/pocketbase/discussions/439#discussioncomment-3651399) and google may reject that for regular google accounts (I have tested it only with google workspace/cloud account, which usually has additional configurations and sometimes is excluded from globally applied google policies).

_In any case, this will be investigated further after the users management refactoring._

**Reply by ghost** [reply]
Well I do have a firewall but the linux logs don't know any blocks being triggered.

My theory is that my systemd file is serving pocketbase as `0.0.0.0` and that may be causing an issue.

**Comment by ganigeorgiev** [comment, marked as answer]
@VentGrey I've just created a regular gmail account and can confirm that it works as expected and the localhost helo/ehlo from https://github.com/pocketbase/pocketbase/discussions/439#discussioncomment-3651399 doesn't seem to cause issues with the test endpooint.

The only specific thing is that you have to enable "App passwords" because "Less secure apps" is no longer available since May 2022:

> To help keep your account secure, from May 30, 2022, ​​Google no longer supports the use of third-party apps or devices which ask you to sign in to your Google Account using only your username and password.
> Important: This deadline does not apply to Google Workspace or Google Cloud Identity customers.

Here are the steps that I've taken:
1. Create a brand new regular gmail account.
2. Enable 2fa (it seems required if you want to use App passwords)
3. Go to https://myaccount.google.com/apppasswords and select the "Other" option from the app dropdown:
    ![screenshot](https://i.imgur.com/VfIMyhn.png).
4. Use the generated password in the smtp settings:
    - host: smtp.gmail.com
    - port: 587
    - tls: auto
    - user: yourusername@gmail.com
    - pass: the generated password from step 3

**Reply by ghost** [reply]
:thinking:  I also followed those steps to no avail :(

**Reply by ghost** [reply]
Both "mail" and "Other" give the same result :(

**Reply by ganigeorgiev** [reply]
@VentGrey Just to make sure that the issue is not related to the Google SMTP configuration, could you try using another SMTP server? I usually test with https://mailtrap.io/ (there is a free plan).

**Reply by ghost** [reply]
I'll try with mailtrap and report back :D 

I just "resetted" firewalld in case that was an issue.

**Reply by ghost** [reply]
Mailtrap also gives a timeout error.
Disabling the firewall completely also doesn't work. I'll try firing up the application without the 0.0.0.0 option and use localhost / 127.0.0.1 instead.

**Reply by ganigeorgiev** [reply]
Mailtrap TLS port is on 587.

Binding to `0.0.0.0` shouldn't be the issue here. I've just tested it on a new Hetzner Ubuntu vm and it works just fine for me.

**Reply by ganigeorgiev** [reply]
@VentGrey I'm not sure how to reproduce the error and I guess it is something related to your network configuration.

Double check the settings and make sure that there are no typos or space characters before/after the configuration fields.

To exclude network misconfiguration, you could also try running PocketBase locally on your machine and try to connect to mailtrap or gmail.

**Reply by ghost** [reply]
Pocketbase locally works. I've checked my configuration three times already but to no avail. Idk if it is possible to log or have a verbose option in pocketbase :(

**Reply by ganigeorgiev** [reply]
You can add the `--debug` flag for more detailed errors and stdout logs, but the timeout connection termination is on the client side and I'm not sure if that be useful.

This most likely is not PocketBase related. You can try sending request to the smtp via `telnet`, eg.:
```
telnet smtp.gmail.com 587

# you may also add an ehlo as an input:
ehlo yourdomain.com
```

If it doesn't work with telnet then there is definelly something else that need to be configured/disabled in your vm.

**Reply by ghost** [reply]
Now that's insteresting...telnet gets stuck. Which is weird since I tried disabling the firewall and it still getting stuck :scream:

**Reply by ghost** [reply]
This is weird since my firewalld configuration seems to be in good condition:

```
firewall-cmd --list-all
public (active)
  target: default
  icmp-block-inversion: no
  interfaces: eth0
  sources: 
  services: dhcpv6-client http https smtp ssh
  ports: 587/tcp
  protocols: 
  forward: no
  masquerade: no
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 
```

**Reply by ganigeorgiev** [reply]
telnet usually wait for user input after connecting, so I'm not sure if that is what you mean by "stuck".

If you are seeing only `Trying IP...` then there is probably something wrong with the network configuration.

A successful telnet response usually will look something like this:

```
Trying IP...
Connected to smtp.gmail.com.
Escape character is '^]'.
220 smtp.gmail.com ESMTP q5-... - gsmtp
... type here the ehlo command ...
```

**Reply by ghost** [reply]
Oh, no it literally gets stuck. Doesn't even reach the `Connected to` part. I've read that some VPS providers block SMTP services unless your acct is old enough or you ask by ticket to prevent users from creating spambots.

**Reply by ganigeorgiev** [reply]
Ah yes, that's also possible. I remember Hetzner disallowing 25 and 465 ports for new accounts.

**Comment by Bunnarin** [comment]
for my case, it was because digital ocean blocked all smtp port
